eclair-core/src/main/scala/fr/acinq/eclair/message/Postman.scala

```diff
@@ -90,55 +83,116 @@ object Postman {
+          Behaviors.same
+        case Subscribe(pathId, replyTo) =>
+          subscribed += (pathId -> replyTo)
+          context.scheduleOnce(nodeParams.onionMessageConfig.timeout, context.self, Unsubscribe(pathId))
+          Behaviors.same
+        case Unsubscribe(pathId) =>
+          subscribed.get(pathId).foreach(ref => {
+            subscribed -= pathId
+            ref ! NoReply
           })
+          Behaviors.same
+      }
+    })
+  }
+}
+
+object SendingMessage {
+  // @formatter:off
+  sealed trait Command
+
+  case object SendMessage extends Command
+  case class SendingStatus(status: MessageRelay.Status) extends Command
+  sealed trait RouteResult extends Command
+  case class MessageRouteFound(messageRoute: Seq[PublicKey]) extends RouteResult
+  case object MessageRouteNotFound extends RouteResult
+  // @formatter:on
+
+  def apply(nodeParams: NodeParams,
+            switchboard: ActorRef[Switchboard.RelayMessage],
+            router: ActorRef[Router.MessageRouteRequest],
+            postman: ActorRef[Postman.Command],
+            destination: Destination,
+            message: TlvStream[OnionMessagePayloadTlv],
+            expectsReply: Boolean,
+            replyTo: ActorRef[Postman.OnionMessageResponse]): Behavior[Command] = {
+    Behaviors.setup(context => {
+      Behaviors.receiveMessagePartial {
+        case SendMessage =>
+          val targetNodeId = destination match {
+            case OnionMessages.BlindedPath(route) => route.introductionNodeId
+            case OnionMessages.Recipient(nodeId, _, _, _) => nodeId
+          }
+          router ! Router.MessageRouteRequest(context.self, nodeParams.nodeId, targetNodeId, Set.empty)
+          waitingForRoute(nodeParams, switchboard, postman, destination, message, expectsReply, replyTo)
+      }
+    })
+  }
+
+  def waitingForRoute(nodeParams: NodeParams,
+                      switchboard: ActorRef[Switchboard.RelayMessage],
+                      postman: ActorRef[Postman.Command],
+                      destination: Destination,
+                      message: TlvStream[OnionMessagePayloadTlv],
+                      expectsReply: Boolean,
+                      replyTo: ActorRef[Postman.OnionMessageResponse]): Behavior[Command] = {
+    Behaviors.setup(context => {
+      Behaviors.receiveMessagePartial {
+        case MessageRouteFound(messageRoute) =>
+          val messageId = randomBytes32()
+          val replyRoute =
+            if (expectsReply) {
+              val numHopsToAdd = 0.max(nodeParams.onionMessageConfig.minIntermediateHops - messageRoute.length)
+              val intermediateHops = (messageRoute.reverse ++ Seq.fill(numHopsToAdd)(nodeParams.nodeId)).map(OnionMessages.IntermediateNode(_))
+              val lastHop = OnionMessages.Recipient(nodeParams.nodeId, Some(messageId))
+              Some(OnionMessages.buildRoute(randomKey(), intermediateHops, lastHop))
+            } else {
+              None
+            }
```

Comment:
This is currently not well-tested at all...can we unit test this in `PostmanSpec`? It's a bit annoying because it requires decryption the onion messages, but I think it's important to have test coverage.

Reply:
You mean testing that the reply route is valid and long enough?

Reply:
Added some tests.